exporter/signalfx/signalfx.go

```diff
@@ -0,0 +1,326 @@
+	}
+	return values
+}
+
+type collector struct {
+	opts Options
+	mu   sync.Mutex // mu guards all the fields.
+
+	// viewData are accumulated and atomically
+	// appended to on every Export invocation, from
+	// stats. These views are cleared out when
+	// Collect is invoked and the cycle is repeated.
+	viewData map[string]*view.Data
+
+	registeredViewsMu sync.Mutex
+
+	registeredViews map[string]string
+}
+
+func (c *collector) addViewData(vd *view.Data) {
+	c.registerViews(vd.View)
+	sig := viewSignature(vd.View.Name, vd.View)
+
+	c.mu.Lock()
+	c.viewData[sig] = vd
+	c.mu.Unlock()
+}
+
+type signalFxMetric struct {
+	metricType  string
+	metricName  string
+	metricValue float64
+	timestamp   time.Time
+	dimensions  map[string]string
+	buckets     []int64
+}
+
+func newCollector(opts Options) *collector {
+	return &collector{
+		opts:            opts,
+		registeredViews: make(map[string]string),
+		viewData:        make(map[string]*view.Data),
+	}
+}
+
+func viewName(namespace string, v *view.View) string {
+	var name string
+	if namespace != "" {
+		name = namespace + "_"
+	}
+	return name + sanitize(v.Name)
+}
+
+func viewSignature(namespace string, v *view.View) string {
+	var buf bytes.Buffer
+	buf.WriteString(viewName(namespace, v))
+	for _, k := range v.TagKeys {
+		buf.WriteString("-" + k.Name())
+	}
+	return buf.String()
+}
+
+// sendRequest sends a package of data containing one metric
+func sendRequest(e *Exporter, data signalFxMetric) {
+	client := sfxclient.NewHTTPSink()
+	client.AuthToken = e.opts.Token
+	ctx := context.Background()
+
+	switch data.metricType {
+	case "gauge":
+		err := client.AddDatapoints(ctx, []*datapoint.Datapoint{
+			sfxclient.GaugeF(data.metricName, data.dimensions, data.metricValue),
+		})
+		if err != nil {
+			e.opts.OnError(errors.New(fmt.Sprintf("Error sending datapoint to SignalFx: %T", err)))
+		}
+	case "cumulative_counter":
+		err := client.AddDatapoints(ctx, []*datapoint.Datapoint{
+			sfxclient.CumulativeF(data.metricName, data.dimensions, data.metricValue),
+		})
+		println("sendind...")
+		if err != nil {
+			e.opts.OnError(errors.New(fmt.Sprintf("Error sending datapoint to SignalFx: %T", err)))
+		}
+	case "cumulative_bucket":
+		bucket := sfxclient.CumulativeBucket{
+			MetricName: data.metricName,
+			Dimensions: data.dimensions,
+		}
+
+		for _, value := range data.buckets {
+			bucket.Add(value)
+		}
+
+		err := client.AddDatapoints(ctx, bucket.Datapoints())
+		if err != nil {
+			e.opts.OnError(errors.New(fmt.Sprintf("Error sending datapoint to SignalFx: %T", err)))
+		}
+	default:
+		e.opts.OnError(errors.New(fmt.Sprintf("Metric type not supported: %s", data.metricType)))
+	}
+}
+
+func (c *collector) cloneViewData() map[string]*view.Data {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	viewDataCopy := make(map[string]*view.Data)
+	for sig, viewData := range c.viewData {
+		viewDataCopy[sig] = viewData
+	}
+	return viewDataCopy
+}
+
+const labelKeySizeLimit = 100
```

Comment:
Our dimension name limit is longer than that (128 I believe). @mdubbyap can you confirm?